### ISSUE TITLE
Database subclasses question

### ISSUE THREAD

**Issue by alemoi** [Starcounter/Starcounter#3053]
Does Starcounter allow adding database classes as subclasses having corresponding SQL support for them, like they are namespaced? Asking because of a good potential isolation for many classes inside one test.

**Comment by k-rus** [comment]
@per-samuelsson was mentioning this case in some issue some time ago. May be he can clarify how weaver deals with this.

**Comment by per-samuelsson** [comment]
Dont understand the question fully? Subclasses are used all the time?

```C#
[Database] public class Foo {}
public class Bar : Foo {}
````

`Bar` will of course be a database class too.

**Comment by k-rus** [comment]
> Does Starcounter allow adding database classes as subclasses having corresponding SQL support for them, like they are namespaced? Asking because of a good potential isolation for many classes inside one test.

Do you mean subclasses or nested classes?

**Comment by miyconst** [comment]
I think @alemoi meant something like this:

```cs
[Database]
public class Order {
	public string Customer;
	public DateTime Date;
	
	[Database]
	public class Item {
		public string Product;
		public int Quantity;
		public decimal Price;
	}
}

var item = new Order.Item();
```

**Comment by per-samuelsson** [comment]
Ah, **nested** classes. That make more sense. :smile: Thanks @miyconst.

We don't support that pattern today, no. I looked in to it once and it turned out the SQL went nuts to resolve that naming pattern.

The weaver fully support it though. You can paste the code provided by @miyconst above to an **app.cs** file and try it yourself:

```
star --sc-compile app.cs
scweaver app.exe
```

Then look at it at some tool, like Reflector. But I doubt you'll be able to load it.

**Comment by per-samuelsson** [comment]
#2375 for some background.

**Comment by k-rus** [comment]
> #2375 for some background.

As I see it should be easy to fix by using dot notation instead of plus sign between classes.

**Comment by per-samuelsson** [comment]
> As I see it should be easy to fix by using dot notation instead of plus sign between classes.

Then maybe we should make an effort to get this to work. @dan31 WDYT?

**Comment by dan31** [comment]
> it turned out the SQL went nuts to resolve that naming pattern.
> As I see it should be easy to fix by using dot notation instead of plus sign between classes.

Especially considering changes in SQL apps level isolation, this should be now a low-hanging fruit (or at least lower). I suggest let's keep it for a moment when we stabilise the new-kernel version of 2.x and then address the issue in a regular manner.

**Comment by alemoi** [comment]
Yeah, nested :) Sorry for late response! Thank you for the answer!

**Comment by bigwad** [comment]
Have spent ten minutes trying to query nested class. Probably, better to report an appropriate error in the weaver until we fix it.
